File: algorithms/2_medium/M/03974_maximum_total_sum_of_k_selected_elements/sort.c
static int cmp(const void* a_ptr, const void* b_ptr)
{
    const int a = *(const int*)a_ptr;
    const int b = *(const int*)b_ptr;
    return (a > b) - (a < b);
}

long long maxSum(int* nums, const int len, const int k, int mul_i32)
{
    assert(len > 0 && k >= 1 && k <= len && mul_i32 >= 1);
    long long mul = mul_i32;
    qsort(nums, (size_t)len, sizeof(*nums), cmp);
    assert(nums[0] >= 1);
    long long sum = 0;

    for (int i = len - 1; i >= len - k; i -= 1) {
        if (mul > 0) {
            sum += mul * nums[i];
            mul -= 1;
        } else {
            sum += nums[i];
        }
    }

    return sum;
}

File: algorithms/2_medium/M/03974_maximum_total_sum_of_k_selected_elements/sort.rs
impl Solution {
    pub fn max_sum(mut nums: Vec<i32>, k: i32, mul: i32) -> i64 {
        assert!(nums.len() > 0 && k >= 1 && (k as usize) <= nums.len() && mul >= 1);
        let k = k as usize;
        let mut mul = mul as i64;
        let len = nums.len();
        nums.sort_unstable();
        assert!(nums[0] >= 1);
        let mut sum = 0_i64;

        for i in ((len - k)..len).rev() {
            if mul > 0 {
                sum += mul * (nums[i] as i64);
                mul -= 1;
            } else {
                sum += nums[i] as i64;
            }
        }

        sum
    }
}
